# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1224781246 14400
#      Thu Oct 23 13:00:46 2008 -0400
# Node ID 095b3e4d64e99065bba14a14fcf849cc68a14fff
# Parent  5a2e45facabe173e147283bbd2cca0864349e4e0
oct-shlib.cc: use RTLD_NOW option for dlopen

diff --git a/liboctave/ChangeLog b/liboctave/ChangeLog
--- a/liboctave/ChangeLog
+++ b/liboctave/ChangeLog
@@ -1,8 +1,13 @@
+2008-10-23  John W. Eaton  <jwe@octave.org>
+
+	* oct-shlib.cc (octave_dlopen_shlib::open): Use RTLD_NOW instead
+	of RTLD_LAZY.
+
 2008-10-12  Jaroslav Hajek <highegg@gmail.com>
 
 	* CSparse.cc (ComplexMatrix::expm): Improve inverse preconditioning
 	according to Marco Caliari.
 	* dSparse.cc (Matrix::expm): Likewise.
 	* fCSparse.cc (FloatComplexMatrix::expm): Likewise.
 	* fSparse.cc (FloatMatrix::expm): Likewise.
 
diff --git a/liboctave/oct-shlib.cc b/liboctave/oct-shlib.cc
--- a/liboctave/oct-shlib.cc
+++ b/liboctave/oct-shlib.cc
@@ -238,18 +238,22 @@ void
 octave_dlopen_shlib::open (const std::string& f)
 {
   if (! is_open ())
     {
       file = f;
 
       int flags = 0;
 
-#if defined (RTLD_LAZY)
-      flags |= RTLD_LAZY;
+      // Use RTLD_NOW to resolve all symbols before dlopen returns.
+      // By using this option, dlopen will detect errors and Octave
+      // won't exit if there are unresolved symbols in the file we are
+      // loading, and we may even get a useful diagnostic.
+#if defined (RTLD_NOW)
+      flags |= RTLD_NOW;
 #endif
 
 #if defined (RTLD_GLOBAL)
       flags |= RTLD_GLOBAL;
 #endif
       
       library = dlopen (file.c_str (), flags);
 
